darshan-util/pydarshan/darshan/experimental/plots/heatmap_handling.py

```diff
@@ -0,0 +1,442 @@
+        # a single one with new row indices
+        write_df = pd.concat(write_df_list, ignore_index=True)
+    else:
+        # if the list is empty assign an empty dataframe
+        write_df = pd.DataFrame()
+
+    return read_df, write_df
+
+
+def get_single_df_dict(
+    report: Any,
+    mods: Sequence[str] = ["DXT_POSIX"],
+    ops: Sequence[str] = ["read", "write"],
+) -> Dict[str, Dict[str, pd.DataFrame]]:
+    """
+    Reorganizes segmented read/write data into a single ``pd.DataFrame``
+    and stores them in a dictionary with an entry for each DXT module.
+
+    Parameters
+    ----------
+
+    report: a ``darshan.DarshanReport``.
+
+    mods: a list of keys designating which Darshan modules to use for
+    data aggregation. Default is ``["DXT_POSIX"]``.
+
+    ops: a list of keys designating which Darshan operations to use for
+    data aggregation. Default is ``["read", "write"]``.
+
+    Returns
+    -------
+
+    flat_data_dict: a nested dictionary where the input module
+    keys (i.e. "DXT_POSIX") are the top level keys, which contain
+    an entry for each input operation (i.e. "read"/"write") that
+    map to dataframes containing all events for the specified operation.
+
+    Examples
+    --------
+    `flat_data_dict` generated from `tests/input/sample-dxt-simple.darshan`:
+        {
+            'DXT_POSIX':
+                {
+                    'read':
+                        Empty DataFrame
+                        Columns: []
+                        Index: [],
+                    'write':
+                        length  start_time  end_time  rank
+                        0      40    0.103379  0.103388     0
+                        1    4000    0.104217  0.104231     0
+                }
+        }
+
+    """
+    # initialize an empty dictionary for storing
+    # module and read/write data
+    flat_data_dict = {}  # type: Dict[str, Dict[str, pd.DataFrame]]
+    # iterate over the modules (i.e. DXT_POSIX)
+    for module_key in mods:
+        # read in the module data, update the name records
+        report.mod_read_all_dxt_records(module_key, dtype="pandas")
+        # retrieve the list of records in pd.DataFrame() form
+        dict_list = report.records[module_key].to_df()
+        # retrieve the list of read/write dataframes from the list of records
+        read_df, write_df = get_rd_wr_dfs(dict_list=dict_list)
+        # create empty dictionary for each module
+        flat_data_dict[module_key] = {}
+        if "read" in ops:
+            # add the concatenated dataframe to the flat dictionary
+            flat_data_dict[module_key]["read"] = read_df
+        if "write" in ops:
+            # add the concatenated dataframe to the flat dictionary
+            flat_data_dict[module_key]["write"] = write_df
+
+    return flat_data_dict
+
+
+def get_aggregate_data(
+    report: Any,
+    mods: Sequence[str] = ["DXT_POSIX"],
+    ops: Sequence[str] = ["read", "write"],
+) -> pd.DataFrame:
```

Comment:
Have you tested the `pandas` types with: https://github.com/pandas-dev/pandas/blob/master/pandas/_typing.py ?

Reply:
We discussed this a bit recently--`pandas` typing isn't really mature/ready for prime time, but it is probably "ok" to leave in the placeholder pandas objects if folks want to, so I'll mark this resolved.